hw/dv/dpi/usbdpi/usbdpi.c

```diff
@@ -650,51 +610,38 @@ void pollRX(struct usbdpi_ctx *ctx, int sendHi, int nakData) {
   }
 }
 
-// Test unimplemented endpoints
-void testUnimplEp(struct usbdpi_ctx *ctx, uint8_t pid) {
+// Test behaviour in (non-)response to other device and unimplemented endpoints
+void testUnimplEp(usbdpi_ctx_t *ctx, uint8_t pid, uint8_t device,
+                  uint8_t endpoint) {
+  usbdpi_transfer_t *tr = ctx->sending;
+  uint8_t *dp;
   switch (ctx->hostSt) {
     case HS_STARTFRAME:
       if ((pid == USB_PID_SETUP) || (pid == USB_PID_OUT)) {
-        ctx->state = ST_SYNC;
-        ctx->bytes = 14;
-        ctx->datastart = 3;
-        ctx->byte = 0;
-        ctx->bit = 1;
-        // The bytes are transmitted LSB to MSB
-        ctx->data[0] = pid;
-        ctx->data[1] =
-            ((UNIMPL_EP_ID & 0x1) << 7) | 0x2;  // endpoint ID LSB, address
-        ctx->data[2] = CRC5((UNIMPL_EP_ID << 7) | 0x2, 11) << 3 |
-                       ((UNIMPL_EP_ID >> 1) & 0x7);  // CRC, endpoint ID MSBs
-
-        ctx->data[3] = USB_PID_DATA0;
-        ctx->data[4] = 0;  // h2d, std, device
-        ctx->data[5] = 5;  // set address
-        ctx->data[6] = 2;  // device address
-        ctx->data[7] = 0;
+        transfer_token(tr, pid, device, endpoint);
+
+        dp = transfer_data_start(tr, USB_PID_DATA0, 0);
+        dp[0] = 0;  // h2d, std, device
+        dp[1] = 5;  // set address
+        dp[2] = 2;  // device address
+        dp[3] = 0;
         // Trigger bitstuffing, technically the device
         // behaviour is unspecified with wIndex != 0
-        ctx->data[8] = 0xFF;  // wIndex = 0xFF00
-        ctx->data[9] = 0;
-        ctx->data[10] = 0;  // wLength = 0
-        ctx->data[11] = 0;
-        add_crc16(ctx->data, ctx->datastart, 12);
+        dp[4] = 0xFF;  // wIndex = 0xFF00
+        dp[5] = 0;
+        dp[6] = 0;  // wLength = 0
+        dp[7] = 0;
+        transfer_data_end(tr, dp + 8);
+
+        transfer_send(ctx, tr);
         ctx->hostSt = HS_WAITACK;
         break;
       } else if (pid == USB_PID_IN) {
-        ctx->state = ST_SYNC;
-        ctx->bytes = 3;
-        ctx->datastart = 0;
-        ctx->byte = 0;
-        ctx->bit = 1;
-        // The bytes are transmitted LSB to MSB
-        ctx->data[0] = pid;
-        ctx->data[1] =
-            ((UNIMPL_EP_ID & 0x1) << 7) | 0x2;  // endpoint ID LSB, address
-        ctx->data[2] = CRC5((UNIMPL_EP_ID << 7) | 0x2, 11) << 3 |
-                       ((UNIMPL_EP_ID >> 1) & 0x7);  // CRC, endpoint ID MSBs
-        // Since the endpoint is not implemented, the device will respond with a
-        // STALL packet (not DATA0/1 or NAK).
+        transfer_token(tr, pid, device, endpoint);
+        transfer_send(ctx, tr);
+
+        // Since the endpoint is not implemented, the device should respond with
+        // a STALL packet (not DATA0/1 or NAK).
```

Comment:
This statement predates your work in this PR, but know that it's wrong and is expected to have been fixed in the actual IP, haha. An unimplemented endpoint cannot respond, since it doesn't exist. Thus, there will be no response.

This expectation didn't cause test failures because the DPI module was not able to check expectations when it was conceived.